pylearn2/datasets/dataset.py

```diff
@@ -48,6 +48,28 @@ def iterator(self, mode=None, batch_size=None, num_batches=None,
             through the dataset and may potentially be shared by
             multiple iterator objects simultaneously (see "Notes"
             below).
+        data_specs : (space, source) pair, optional
+            `space` is an instance of `Space` (possibly a
+            `CompositeSpace`), and `source` is a string (or tuple of
+            strings if `space` is a `CompositeSpace`). `source` defines
+            which data should be returned by the iterator (for instance
+            'features', 'targets'), and `space` specifies in which
+            format it will be returned. These specifications should be
+            flat: if `space` is a `CompositeSpace`, the sub-spaces should
```

Comment:
I am not aware of any contexts where a provided `Space` _must_ be a `CompositeSpace`. There are however contexts where all spaces created or _returned_ by a function will be instances of `CompositeSpace`, even if that is not required by the caller of that function.
I don't necessarily see a problem in being stricter in the value returned at these places, but more tolerant in the values used by `Dataset.iterator`, especially since the `iterator` method can sometimes be called from a user's script.